### DISCUSSION TITLE
variable radius with depth feature request in torch_cluster.radius_cuda

### DISCUSSION THREAD

**Discussion by imransai** [pyg-team/pytorch_geometric#3007] [Ideas]
Hi Matthias,
I was working on an algorithm which would collect all pixels within in a radius that varies with depth/range. Your radius () seems to accept a fixed radius to collect neighbors within a certain 3D point. But I am not sure how to implement a variable radius with cuda. Is the variable radius idea easier to integrate? Or can you show me how to modify your code to make it variable radius implementation in cuda?

Let me know if you need more info. Really waiting for your reply. :)

**Comment by rusty1s** [comment]
Instead of passing a fixed scalar to `radius`, you would need to pass in a vector, holding the radius for each point. The radius check then happens [on a per point basis](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cuda/radius_cuda.cu#L34).

**Reply by imransai** [reply]
I see that makes sense. I guess that is the only .cu file that needs to be changed? The question is how to integrate it as a python wrapper for use with python? I am newbie in wrapping up a cuda code in python module. Is there any practice or suggestions you would like me to follow?

**Reply by rusty1s** [reply]
The simplest way would be to fork `torch-sparse` and modify the code accordingly, i.e., `csrc/radius.cpp`, `csrc/cuda/radius_cuda.h` and `csrc/cuda/radius_cuda.cu` now take in a `torch::Tensor` for `r` instead of a single scalar. Then, in `csrc/cuda/radius_cuda.cu`, you need to pass in `r.data_ptr<scalar_t>()` to `radius_kernel`. In the `radius_kernel`, you simply access `radius[n_y]` instead of `radius` [here](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cuda/radius_cuda.cu#L34).

The Python bindings are handled [here](https://github.com/rusty1s/pytorch_cluster/blob/master/torch_cluster/radius.py), in which `r` should be replaced by a `torch.Tensor` as well.

Let me know if this information helps :)
